querqy-solr tests sometimes fail due to random locale

Using the Solr testing framework, tests sometimes fail. It seems that this is triggered by some random locale chosen by the framework. One such locale is `tr-CY`, so 

`mvn test -Dtests.locale=tr-CY`

always fails. Example error:

```
[ERROR] testThatOptCanBePassedAsBoostMethodParam(querqy.solr.BoostMethodTest)  Time elapsed: 0.002 s  <<< ERROR!
java.lang.RuntimeException: org.apache.solr.core.SolrCoreInitializationException: SolrCore 'collection1' is not available due to init failure: querqy.rewrite.commonrules.RuleParseException: Line 2: @ expected at beginning of line: FILTER: * f2:c
	at querqy.solr.BoostMethodTest.setUp(BoostMethodTest.java:37)
Caused by: org.apache.solr.core.SolrCoreInitializationException: SolrCore 'collection1' is not available due to init failure: querqy.rewrite.commonrules.RuleParseException: Line 2: @ expected at beginning of line: FILTER: * f2:c
	at querqy.solr.BoostMethodTest.setUp(BoostMethodTest.java:37)
```

--- a/querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java
+++ b/querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java
@@ -26,7 +26,8 @@ public class LineParser {
 
     static final char RAWQUERY = '*';
 
-    public static Object parse(String line, Input previousInput, QuerqyParserFactory querqyParserFactory) {
+    public static Object parse(final String line, final Input previousInput,
+                               final QuerqyParserFactory querqyParserFactory) {
 
 
         if (line.endsWith("=>")) {
@@ -40,7 +41,7 @@ public class LineParser {
             return new ValidationError("Missing input for instruction");
         }
 
-        String lcLine = line.toLowerCase().trim();
+        final String lcLine = line.toLowerCase(Locale.ROOT).trim();
 
         if (lcLine.startsWith(INSTR_DELETE)) {
 
@@ -58,14 +59,14 @@ public class LineParser {
             }
 
             instructionTerms = instructionTerms.substring(1).trim();
-            Object expr = parseTermExpression(instructionTerms);
+            final Object expr = parseTermExpression(instructionTerms);
             if (expr instanceof ValidationError) {
                 return new ValidationError("Cannot parse line: " + line + " : " + ((ValidationError) expr).getMessage());
             }
             @SuppressWarnings("unchecked")
-            List<Term> deleteTerms = (List<Term>) expr;
-            List<Term> inputTerms = previousInput.getInputTerms();
-            for (Term term : deleteTerms) {
+            final List<Term> deleteTerms = (List<Term>) expr;
+            final List<Term> inputTerms = previousInput.getInputTerms();
+            for (final Term term : deleteTerms) {
                 if (term.findFirstMatch(inputTerms) == null) {
                     return new ValidationError("Condition doesn't contain the term to delete: " + term);
                 }

--- a/querqy-core/src/test/java/querqy/rewrite/commonrules/LineParserTest.java
+++ b/querqy-core/src/test/java/querqy/rewrite/commonrules/LineParserTest.java
@@ -5,17 +5,56 @@ import static org.hamcrest.collection.IsIterableContainingInAnyOrder.containsInA
 import static org.hamcrest.Matchers.*;
 
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Locale;
 
 import org.hamcrest.Description;
 import org.hamcrest.TypeSafeMatcher;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
 
 import querqy.rewrite.commonrules.model.*;
 import querqy.rewrite.commonrules.model.BoostInstruction.BoostDirection;
 
 public class LineParserTest {
+
+    private Locale locale;
+
+    @Before
+    public void saveDefaultLocale() {
+        locale = Locale.getDefault();
+    }
+
+    @After
+    public void restoreDefaultLocale() {
+        Locale.setDefault(locale);
+    }
+
+    @Test
+    public void testPredicatesWithVaryingLocales() {
+
+        final Input input = new Input(Collections.singletonList(new Term("a".toCharArray(), 0, 1, null)), "a");
+        final WhiteSpaceQuerqyParserFactory rhsParserFactory = new WhiteSpaceQuerqyParserFactory();
+
+        for (final Locale locale: Arrays.asList(Locale.ENGLISH, new Locale("tr", "CY"))) {
+
+            Locale.setDefault(locale);
+
+            assertTrue(LineParser.parse("filter: f", input, rhsParserFactory) instanceof FilterInstruction);
+            assertTrue(LineParser.parse("FILTER: f", input, rhsParserFactory) instanceof FilterInstruction);
+            assertTrue(LineParser.parse("up: f", input, rhsParserFactory) instanceof BoostInstruction);
+            assertTrue(LineParser.parse("UP: f", input, rhsParserFactory) instanceof BoostInstruction);
+            assertTrue(LineParser.parse("down: f", input, rhsParserFactory) instanceof BoostInstruction);
+            assertTrue(LineParser.parse("DOWN: f", input, rhsParserFactory) instanceof BoostInstruction);
+            assertTrue(LineParser.parse("delete: a", input, rhsParserFactory) instanceof DeleteInstruction);
+            assertTrue(LineParser.parse("DELETE: a", input, rhsParserFactory) instanceof DeleteInstruction);
+
+        }
+
+    }
 
     @Test
     public void testParseTermValueOnly() {
@@ -27,7 +66,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testParseSingleLetterValue() throws Exception {
+    public void testParseSingleLetterValue() {
         Term term = LineParser.parseTerm("a");
         assertEquals(1, term.length());
         assertArrayEquals(new char[] {'a'},new char[] {term.charAt(0)});
@@ -36,25 +75,25 @@ public class LineParserTest {
     }
     
     @Test
-    public void testParseTermWithFieldName() throws Exception {
+    public void testParseTermWithFieldName() {
         Term term = LineParser.parseTerm("f1:abc");
         assertEquals(3, term.length());
         assertArrayEquals(new char[] {'a', 'b', 'c'},new char[] {term.charAt(0), term.charAt(1), term.charAt(2)});
         assertFalse(term instanceof PrefixTerm);
-        assertEquals(Arrays.asList("f1"), term.getFieldNames());
+        assertEquals(Collections.singletonList("f1"), term.getFieldNames());
     }
     
     @Test
-    public void testParseSingleLetterValueWithFieldName() throws Exception {
+    public void testParseSingleLetterValueWithFieldName() {
         Term term = LineParser.parseTerm("f1:a");
         assertEquals(1, term.length());
         assertArrayEquals(new char[] {'a'},new char[] {term.charAt(0)});
         assertFalse(term instanceof PrefixTerm);
-        assertEquals(Arrays.asList("f1"), term.getFieldNames());
+        assertEquals(Collections.singletonList("f1"), term.getFieldNames());
     }
     
     @Test
-    public void testParseTermWithFieldNames() throws Exception {
+    public void testParseTermWithFieldNames() {
         Term term = LineParser.parseTerm("{f1,f2}:abc");
         assertEquals(3, term.length());
         assertArrayEquals(new char[] {'a', 'b', 'c'},new char[] {term.charAt(0), term.charAt(1), term.charAt(2)});
@@ -63,7 +102,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testParseTermWithFieldNamesContainingSpace() throws Exception {
+    public void testParseTermWithFieldNamesContainingSpace() {
         assertThat(LineParser.parseTerm("{ f1 , f2 }:abc"), term("abc", "f1", "f2"));
     }
     
@@ -77,22 +116,22 @@ public class LineParserTest {
     }
     
     @Test
-    public void testParseSingleLetterPrefix() throws Exception {
+    public void testParseSingleLetterPrefix() {
         assertThat(LineParser.parseTerm("a*"), prefix("a"));
     }
     
     @Test
-    public void testParsePrefixWithFieldName() throws Exception {
+    public void testParsePrefixWithFieldName() {
         assertThat(LineParser.parseTerm("f1:abc*"), prefix("abc", "f1"));
     }
     
     @Test
-    public void testParsePrefixWithFieldNames() throws Exception {
+    public void testParsePrefixWithFieldNames() {
         assertThat(LineParser.parseTerm("{f1,f2}:abc*"), prefix("abc", "f1", "f2"));
     }
     
     @Test
-    public void testThatWildcardOnlyTermIsNotAllowed() throws Exception {
+    public void testThatWildcardOnlyTermIsNotAllowed() {
         try {
             LineParser.parseTerm("*");
             fail("Wildcard-only term must not be allowed");
@@ -102,7 +141,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatWildcardOnlyTermIsNotAllowedWithFieldName() throws Exception {
+    public void testThatWildcardOnlyTermIsNotAllowedWithFieldName() {
         try {
             LineParser.parseTerm("f1:*");
             fail("Wildcard-only term must not be allowed with fieldname");
@@ -112,7 +151,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatWildcardOnlyTermIsNotAllowedWithFieldNames() throws Exception {
+    public void testThatWildcardOnlyTermIsNotAllowedWithFieldNames() {
         try {
             LineParser.parseTerm("{f1,f2}:*");
             fail("Wildcard-only term must not be allowed with fieldname");
@@ -122,14 +161,14 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatWildcardCannotBeFollowedByRightBoundary() throws Exception {
+    public void testThatWildcardCannotBeFollowedByRightBoundary() {
         Object parseResult = LineParser.parseInput("a" + LineParser.WILDCARD + LineParser.BOUNDARY);
         assertEquals("Wildcard should not be allowed before right boundary", 
                 new ValidationError(LineParser.WILDCARD + " cannot be combined with right boundary"), parseResult);
     }
     
     @Test
-    public void testThatWildcardCanBeCombinedWithLeftBoundary() throws Exception {
+    public void testThatWildcardCanBeCombinedWithLeftBoundary() {
         Object parseResult = LineParser.parseInput(LineParser.BOUNDARY + "a" + LineParser.WILDCARD);
         assertTrue(parseResult instanceof Input);
         Input input = (Input) parseResult;
@@ -138,7 +177,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatBoundariesAreParsedInInput() throws Exception {
+    public void testThatBoundariesAreParsedInInput() {
         Object parseResult = LineParser.parseInput(LineParser.BOUNDARY + "a" + LineParser.BOUNDARY);
         assertTrue(parseResult instanceof Input);
         Input input = (Input) parseResult;
@@ -147,7 +186,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatBoundariesAreParsedInOtherwiseEmptyInput() throws Exception {
+    public void testThatBoundariesAreParsedInOtherwiseEmptyInput() {
         Object parseResult = LineParser.parseInput(LineParser.BOUNDARY + "" + LineParser.BOUNDARY);
         assertTrue(parseResult instanceof Input);
         Input input = (Input) parseResult;
@@ -156,7 +195,7 @@ public class LineParserTest {
     }
 
     @Test
-    public void testThatBoostInstructionWithSingleLetterTermIsAccepted() throws Exception {
+    public void testThatBoostInstructionWithSingleLetterTermIsAccepted() {
         String line = "UP: x";
         String lcLine = line.toLowerCase();
         final Object instruction = LineParser
@@ -165,7 +204,7 @@ public class LineParserTest {
     }
 
     @Test
-    public void testThatBoostInstructionWithSingleLetterTermAndBoostFactorIsAccepted() throws Exception {
+    public void testThatBoostInstructionWithSingleLetterTermAndBoostFactorIsAccepted() {
         String line = "UP(5): x";
         String lcLine = line.toLowerCase();
         final Object instruction = LineParser
@@ -174,7 +213,7 @@ public class LineParserTest {
     }
 
     @Test
-    public void testThatPlayholdersAreParsedForBoostInstruction() throws Exception {
+    public void testThatPlayholdersAreParsedForBoostInstruction() {
         String line = "UP(500): 3$1";
         String lcLine = line.toLowerCase();
         final Object instruction = LineParser
@@ -229,7 +268,7 @@ public class LineParserTest {
     }
     
     @Test
-    public void testThatCaseIsPreservedInDecorateInstruction() throws Exception {
+    public void testThatCaseIsPreservedInDecorateInstruction() {
         Input input = (Input) LineParser.parseInput("in");
         assertEquals(new DecorateInstruction("Some Deco"), LineParser.parse("DECORATE: Some Deco", input, null));
     }

--- a/querqy-for-lucene/pom.xml
+++ b/querqy-for-lucene/pom.xml
@@ -73,7 +73,7 @@
         <hamcrest.version>2.1</hamcrest.version>
         <mockito.version>2.27.0</mockito.version>
 
-        <querqy.core.version>3.5.1</querqy.core.version>
+        <querqy.core.version>3.6.0-SNAPSHOT</querqy.core.version>
         <lucene.version>8.1.0</lucene.version>
         <minidev.json-smart.version>2.3</minidev.json-smart.version>
         <commons.io.version>2.5</commons.io.version>

--- a/querqy-for-lucene/querqy-solr/src/test/java/querqy/solr/CriteriaSelectionDefaultsTest.java
+++ b/querqy-for-lucene/querqy-solr/src/test/java/querqy/solr/CriteriaSelectionDefaultsTest.java
@@ -222,7 +222,7 @@ public class CriteriaSelectionDefaultsTest extends SolrTestCaseJ4 {
         SolrQueryRequest req = req("q", "input5 input6",
                 DisMaxParams.QF, "f1",
                 DisMaxParams.MM, "1",
-                getFilterParamName(REWRITER_ID_3), "$[?('a' in @.tt)]",
+                getFilterParamName(REWRITER_ID_3), "$[?('a' IN @.tt)]",
                 "defType", "querqy",
                 "debugQuery", "true"
         );

--- a/querqy-for-lucene/querqy-solr/src/test/java/querqy/solr/CriteriaSelectionTest.java
+++ b/querqy-for-lucene/querqy-solr/src/test/java/querqy/solr/CriteriaSelectionTest.java
@@ -240,7 +240,7 @@ public class CriteriaSelectionTest extends SolrTestCaseJ4 {
                 DisMaxParams.QF, "f1",
                 DisMaxParams.MM, "1",
                 getStrategyParamName(REWRITER_ID_3), "criteria",
-                getFilterParamName(REWRITER_ID_3), "$[?('a' in @.tt)]",
+                getFilterParamName(REWRITER_ID_3), "$[?('a' IN @.tt)]",
                 "defType", "querqy",
                 "debugQuery", "true"
         );

